### PR TITLE
When CUDA libs cannot be loaded, TF should exit

### PR DESCRIPTION
**Pull request by burgerkingeater**
When a job cannot acquire GPU due to some CUDA libs missing, TF should fail early instead of falling back to CPU.  
If TF ignores the error and continue, it could potentially cause unexpected behavior in MWMS - E.g one worker has GPU while one doesn't. Currently when MWMS is running with mixed setup like this, it just hung.

Based on the discussion where the original change of falling back to CPU when CUDA lib cannot be loaded(https://github.com/tensorflow/tensorflow/commit/fcfb99bc3aa3110feae5233dbe4634bffa70dc59) was originated from: https://groups.google.com/a/tensorflow.org/g/developers/c/MUq0ee3g-xk/m/4H84DeiAAQAJ , it mentioned two options facing CUDA lib not being installed properly
 1. fallback to CPU
 2. exit with error message.

option 2 makes more sense to me as I stated above.

### CODE REVIEW

**Comment by burgerkingeater** [comment]
@rohan100jain would you mind taking a look?

**Comment by rohan100jain** [comment]
@sanjoy @mihaimaruseac can you take a look?

**Comment by mihaimaruseac** [comment]
There's also the issue that users installing `tensorflow` on systems with no GPU will get a broken package if this lands. Any thing they try to do on TF would result in an exit.

Should we instead control this via an enviroment variable / config flag?

**Comment by burgerkingeater** [comment]
@mihaimaruseac in this case, TF won't even invoke this gpu_device.cc. It will only be invoked when gpu is available and required by user.

**Comment by mihaimaruseac** [comment]
Oh, thank you, then I guess we can take this in and run some testing after it gets into tf-nightly to prevent unwanted regression

**Review by mihaimaruseac — APPROVED** [review]

**Comment by burgerkingeater** [comment]
@mihaimaruseac sounds good, thanks!

**Comment by burgerkingeater** [comment]
> @sanjoy @mihaimaruseac can you take a look? This might break a lot of users but might be a clearer UX. One naive question is that do we distinguish between folks having an NVIDIA driver but the wrong CUDA setup vs folks not having any NVIDIA gpus at all?

To answer your question, yes, we distinguish these two cases - the gpu_device.cc is invoked only when gpu is available to the tf process.

**Comment by burgerkingeater** [comment]
@mihaimaruseac it seems build is failing due to some unrelated test cases failing. Any idea on this?

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 1]
```diff
@@ -1767,9 +1767,8 @@ Status BaseGPUDeviceFactory::GetValidDeviceIds(
                     "if you would like to use GPU. Follow the guide at "
                     "https://www.tensorflow.org/install/gpu for how to "
                     "download and setup the required libraries for your "
-                    "platform.\nSkipping registering "
-                    "GPU devices...";
-    return Status::OK();
+                    "platform.\n";
+    return errors::FailedPrecondition("Cannot dlopen some GPU libraries.");
```
Comment (sanjoy) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc:1771]: Why not return the failed `Status` in `handle_or`?

**Review by sanjoy — CHANGES_REQUESTED** [review]
This looks fine overall, but let's land this (immediately) after TF 2.6 is cut so that users have time to complain if it breaks them.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc (hunk 1 above)
Comment (burgerkingeater) [reply tensorflow/core/common_runtime/gpu/gpu_device.cc]: thanks, updated.

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by burgerkingeater** [comment]
@gbaned how about this PR? anything I can help to get it merged?

**Comment by PatriceVignola** [comment]
> This looks fine overall, but let's land this (immediately) after TF 2.6 is cut so that users have time to complain if it breaks them.

Just an FYI that this breaks pluggable devices that were previously depending on the `tf-nightly` package when CUDA libraires are not installed. This is not blocking since people can use `tf-nightly-cpu` instead, but the [pluggable devices tutorial](https://github.com/tensorflow/community/pull/352/files) says to install the former which can be confusing.

**Comment by sanjoy** [comment]
> @sanjoy Just an FYI that this breaks pluggable devices that were previously depending on the `tf-nightly` package when CUDA libraires are not installed. This is not blocking since people can use `tf-nightly-cpu` instead, but the [pluggable devices tutorial](https://github.com/tensorflow/community/pull/352/files) says to install the former which can be confusing.

Thanks, looks like being cautious here w.r.t. TF 2.6 paid off. :)

CC @penpornk to advise on PluggableDevice.

**Comment by penpornk** [comment]
@PatriceVignola That part of the tutorial was written before PluggableDevice landed in TF 2.5. Now the users can just install TF 2.6 (or 2.5) instead of the nightly packages. We'll update the tutorial. 

> Thanks, looks like being cautious here w.r.t. TF 2.6 paid off. :)
Thank you for landing this after 2.6, @sanjoy! :)

**Comment by mihaimaruseac** [comment]
I actually think we should roll this back.

Now users that install tensorflow (tf-nightly for now, but from TF 2.7 where this PR exists) won't be able to use it if they don't have CUDA libs installed:

```
(venv) mihaimaruseac@ankh:/tmp/tff$ python
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
2021-07-28 12:57:55.695347: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-07-28 12:57:55.695388: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
>>> tf.constant([1,2,3])
2021-07-28 12:58:18.836570: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.836700: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublas.so.11'; dlerror: libcublas.so.11: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.836808: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcublasLt.so.11'; dlerror: libcublasLt.so.11: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.836905: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.837002: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.837098: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.837193: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusparse.so.11'; dlerror: libcusparse.so.11: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.837297: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open shared object file: No such file or directory
2021-07-28 12:58:18.837321: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1829] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py", line 271, in constant
    return _constant_impl(value, dtype, shape, name, verify_shape=False,
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py", line 283, in _constant_impl
    return _constant_eager_impl(ctx, value, dtype, shape, verify_shape)
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py", line 308, in _constant_eager_impl
    t = convert_to_eager_tensor(value, ctx, dtype)
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/framework/constant_op.py", line 105, in convert_to_eager_tensor
    ctx.ensure_initialized()
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/eager/context.py", line 538, in ensure_initialized
    config_str = self.config.SerializeToString()
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/eager/context.py", line 987, in config
    self._initialize_physical_devices()
  File "/tmp/tff/venv/lib/python3.9/site-packages/tensorflow/python/eager/context.py", line 1320, in _initialize_physical_devices
    devs = pywrap_tfe.TF_ListPhysicalDevices()
tensorflow.python.framework.errors_impl.InternalError: Cannot dlopen all CUDA libraries.
```

Note, there is no explicit call of a CUDA operation. We should just fallback to the CPU implementation here.

**Comment by burgerkingeater** [comment]
@mihaimaruseac thanks, I will work on this today.

**Comment by GMNGeoffrey** [comment]
Note that there are also reasons to want GPU and not CUDA. This broke all of our vulkan tests when we tried to upgrade our TF pip package.

**Pull request merged**